compile error on rawhide

Please fix on `stable` if possible.
```
2021-01-05T14:56:25.6383348Z In file included from /__w/xtp/xtp/votca/xtp/include/votca/xtp/bfgs-trm.h:26,
2021-01-05T14:56:25.6384375Z                  from /__w/xtp/xtp/votca/xtp/src/libxtp/bfgs-trm.cc:22:
2021-01-05T14:56:25.6386359Z /__w/xtp/xtp/votca/xtp/include/votca/xtp/logger.h: In member function 'votca::xtp::Logger& votca::xtp::Logger::operator()(votca::Log::Level)':
2021-01-05T14:56:25.6387703Z /__w/xtp/xtp/votca/xtp/include/votca/xtp/logger.h:182:61: error: 'this' pointer null [-Werror=nonnull]
2021-01-05T14:56:25.6388778Z   182 |     dynamic_cast<LogBuffer *>(rdbuf())->setLogLevel(LogLevel);
2021-01-05T14:56:25.6389495Z       |                                                             ^
2021-01-05T14:56:25.6439748Z /__w/xtp/xtp/votca/xtp/include/votca/xtp/logger.h:51:8: note: in a call to non-static member function 'void votca::xtp::LogBuffer::setLogLevel(votca::Log::Level)'
2021-01-05T14:56:25.6441034Z    51 |   void setLogLevel(Log::Level LogLevel) { _LogLevel = LogLevel; }
2021-01-05T14:56:25.6441617Z       |        ^~~~~~~~~~~
2021-01-05T14:56:25.6447746Z /__w/xtp/xtp/votca/xtp/include/votca/xtp/logger.h: In member function 'void votca::xtp::Logger::setMultithreading(bool)':
2021-01-05T14:56:25.6449092Z /__w/xtp/xtp/votca/xtp/include/votca/xtp/logger.h:189:68: error: 'this' pointer null [-Werror=nonnull]
2021-01-05T14:56:25.6450305Z   189 |     dynamic_cast<LogBuffer *>(rdbuf())->setMultithreading(_maverick);
2021-01-05T14:56:25.6450958Z       |                                                                    ^
2021-01-05T14:56:25.6452687Z /__w/xtp/xtp/votca/xtp/include/votca/xtp/logger.h:54:8: note: in a call to non-static member function 'void votca::xtp::LogBuffer::setMultithreading(bool)'
2021-01-05T14:56:25.6453844Z    54 |   void setMultithreading(bool maverick) { _maverick = maverick; }
2021-01-05T14:56:25.6454465Z       |        ^~~~~~~~~~~~~~~~~
2021-01-05T14:56:25.6460472Z /__w/xtp/xtp/votca/xtp/include/votca/xtp/logger.h: In member function 'void votca::xtp::Logger::setPreface(votca::Log::Level, const string&)':
2021-01-05T14:56:25.6461917Z /__w/xtp/xtp/votca/xtp/include/votca/xtp/logger.h:196:66: error: 'this' pointer null [-Werror=nonnull]
2021-01-05T14:56:25.6463019Z   196 |     dynamic_cast<LogBuffer *>(rdbuf())->setPreface(level, preface);
2021-01-05T14:56:25.6463664Z       |                                                                  ^
2021-01-05T14:56:25.6465187Z /__w/xtp/xtp/votca/xtp/include/votca/xtp/logger.h:57:8: note: in a call to non-static member function 'void votca::xtp::LogBuffer::setPreface(votca::Log::Level, std::string)'
2021-01-05T14:56:25.6466366Z    57 |   void setPreface(Log::Level level, std::string preface) {
2021-01-05T14:56:25.6466941Z       |        ^~~~~~~~~~
2021-01-05T14:56:25.6517673Z /__w/xtp/xtp/votca/xtp/include/votca/xtp/logger.h: In member function 'void votca::xtp::Logger::EnablePreface()':
2021-01-05T14:56:25.6518996Z /__w/xtp/xtp/votca/xtp/include/votca/xtp/logger.h:206:76: error: 'this' pointer null [-Werror=nonnull]
2021-01-05T14:56:25.6520180Z   206 |   void EnablePreface() { dynamic_cast<LogBuffer *>(rdbuf())->EnablePreface(); }
2021-01-05T14:56:25.6520885Z       |                                                                            ^
2021-01-05T14:56:25.6522337Z /__w/xtp/xtp/votca/xtp/include/votca/xtp/logger.h:75:8: note: in a call to non-static member function 'void votca::xtp::LogBuffer::EnablePreface()'
2021-01-05T14:56:25.6523387Z    75 |   void EnablePreface() { _writePreface = true; }
2021-01-05T14:56:25.6523918Z       |        ^~~~~~~~~~~~~
2021-01-05T14:56:25.6525137Z /__w/xtp/xtp/votca/xtp/include/votca/xtp/logger.h: In member function 'void votca::xtp::Logger::DisablePreface()':
2021-01-05T14:56:25.6526428Z /__w/xtp/xtp/votca/xtp/include/votca/xtp/logger.h:209:56: error: 'this' pointer null [-Werror=nonnull]
2021-01-05T14:56:25.6527477Z   209 |     dynamic_cast<LogBuffer *>(rdbuf())->DisablePreface();
2021-01-05T14:56:25.6528070Z       |                                                        ^
2021-01-05T14:56:25.6529489Z /__w/xtp/xtp/votca/xtp/include/votca/xtp/logger.h:76:8: note: in a call to non-static member function 'void votca::xtp::LogBuffer::DisablePreface()'
2021-01-05T14:56:25.6530565Z    76 |   void DisablePreface() { _writePreface = false; }
2021-01-05T14:56:25.6531348Z       |        ^~~~~~~~~~~~~~
2021-01-05T14:56:25.6532567Z /__w/xtp/xtp/votca/xtp/include/votca/xtp/logger.h: In member function 'std::string votca::xtp::Logger::Messages()':
2021-01-05T14:56:25.6533808Z /__w/xtp/xtp/votca/xtp/include/votca/xtp/logger.h:220:57: error: 'this' pointer null [-Werror=nonnull]
2021-01-05T14:56:25.6534967Z   220 |     return dynamic_cast<LogBuffer *>(rdbuf())->Messages();
2021-01-05T14:56:25.6535519Z       |                                                         ^
2021-01-05T14:56:25.6536930Z /__w/xtp/xtp/votca/xtp/include/votca/xtp/logger.h:85:15: note: in a call to non-static member function 'std::string votca::xtp::LogBuffer::Messages()'
2021-01-05T14:56:25.6537838Z    85 |   std::string Messages() {
2021-01-05T14:56:25.6538258Z       |               ^~~~~~~~
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -3,6 +3,7 @@ For more detailed information about the changes see the history of the
 
 ## Version 1.6.4 (released XX.12.20)
 * fix build on openSUSE (#622)
+* Refactored logger (#625)
 
 ## Version 1.6.3 (released 09.12.20)
 * switch to ghcr.io for CI (#555)

--- a/include/votca/xtp/logger.h
+++ b/include/votca/xtp/logger.h
@@ -42,7 +42,7 @@ namespace xtp {
 /*
  * Custom buffer to store messages
  */
-class LogBuffer : public std::stringbuf {
+class LogBuffer final : public std::stringbuf {
 
  public:
   LogBuffer() : std::stringbuf() { _LogLevel = Log::current_level; }
@@ -105,7 +105,7 @@ class LogBuffer : public std::stringbuf {
   bool _writePreface = true;
 
  protected:
-  int sync() override {
+  int sync() {
 
     std::ostringstream _message;
 
@@ -165,35 +165,30 @@ class Logger final : public std::ostream {
   }
 
  public:
-  Logger() : std::ostream(new LogBuffer()), _ReportLevel(Log::current_level) {
+  Logger() : std::ostream(&_buffer), _ReportLevel(Log::current_level) {
     setMultithreading(_maverick);
   }
   Logger(Log::Level ReportLevel)
-      : std::ostream(new LogBuffer()), _ReportLevel(ReportLevel) {
+      : std::ostream(&_buffer), _ReportLevel(ReportLevel) {
     setMultithreading(_maverick);
   }
 
-  ~Logger() final {
-    delete rdbuf();
-    rdbuf(nullptr);
-  }
-
   Logger &operator()(Log::Level LogLevel) {
-    dynamic_cast<LogBuffer *>(rdbuf())->setLogLevel(LogLevel);
+    _buffer.setLogLevel(LogLevel);
     return *this;
   }
 
   void setReportLevel(Log::Level ReportLevel) { _ReportLevel = ReportLevel; }
   void setMultithreading(bool maverick) {
     _maverick = maverick;
-    dynamic_cast<LogBuffer *>(rdbuf())->setMultithreading(_maverick);
+    _buffer.setMultithreading(_maverick);
   }
   bool isMaverick() const { return _maverick; }
 
   Log::Level getReportLevel() const { return _ReportLevel; }
 
   void setPreface(Log::Level level, const std::string &preface) {
-    dynamic_cast<LogBuffer *>(rdbuf())->setPreface(level, preface);
+    _buffer.setPreface(level, preface);
   }
 
   void setCommonPreface(const std::string &preface) {
@@ -203,22 +198,19 @@ class Logger final : public std::ostream {
     setPreface(Log::warning, preface);
   }
 
-  void EnablePreface() { dynamic_cast<LogBuffer *>(rdbuf())->EnablePreface(); }
+  void EnablePreface() { _buffer.EnablePreface(); }
 
-  void DisablePreface() {
-    dynamic_cast<LogBuffer *>(rdbuf())->DisablePreface();
-  }
+  void DisablePreface() { _buffer.DisablePreface(); }
 
  private:
+  LogBuffer _buffer;
   // at what level of detail output messages
   Log::Level _ReportLevel = Log::error;
 
   // if true, only a single processor job is executed
   bool _maverick = false;
 
-  std::string Messages() {
-    return dynamic_cast<LogBuffer *>(rdbuf())->Messages();
-  }
+  std::string Messages() { return _buffer.Messages(); }
 };
 
 /**

